Compilation of CONFIG_NANO fails

**Description**
Compiling Git master with ```CONFIG_NANO``` fails.

```
make[3]: Entering directory '/home/elks/Downloads/elks/elks/net'
ia16-elf-gcc -ffreestanding -fno-inline -melks -mcmodel=small -msegment-relocation-stuff -mtune=i8086  -I/home/elks/Downloads/elks/include -I/home/elks/Downloads/elks/elks/include -I/home/elks/Downloads/elks/libc/include -DUTS_RELEASE=\"0.4.0\" -D__KERNEL__ -Wall -Os -Wno-strict-aliasing -fno-delete-null-pointer-checks -Wnull-dereference -c -o socket.o socket.c
ia16-elf-gcc -ffreestanding -fno-inline -melks -mcmodel=small -msegment-relocation-stuff -mtune=i8086  -I/home/elks/Downloads/elks/include -I/home/elks/Downloads/elks/elks/include -I/home/elks/Downloads/elks/libc/include -DUTS_RELEASE=\"0.4.0\" -D__KERNEL__ -Wall -Os -Wno-strict-aliasing -fno-delete-null-pointer-checks -Wnull-dereference -c -o protocols.o protocols.c
ia16-elf-gcc -ffreestanding -fno-inline -melks -mcmodel=small -msegment-relocation-stuff -mtune=i8086  -I/home/elks/Downloads/elks/include -I/home/elks/Downloads/elks/elks/include -I/home/elks/Downloads/elks/libc/include -DUTS_RELEASE=\"0.4.0\" -D__KERNEL__ -Wall -Os -Wno-strict-aliasing -fno-delete-null-pointer-checks -Wnull-dereference -c -o ipv4/af_inet.o ipv4/af_inet.c
ia16-elf-gcc -ffreestanding -fno-inline -melks -mcmodel=small -msegment-relocation-stuff -mtune=i8086  -I/home/elks/Downloads/elks/include -I/home/elks/Downloads/elks/elks/include -I/home/elks/Downloads/elks/libc/include -DUTS_RELEASE=\"0.4.0\" -D__KERNEL__ -Wall -Os -Wno-strict-aliasing -fno-delete-null-pointer-checks -Wnull-dereference -c -o nano/af_nano.o nano/af_nano.c
nano/af_nano.c: In function 'nano_bind':
nano/af_nano.c:152:20: warning: unused variable 'old_ds' [-Wunused-variable]
     unsigned short old_ds;
                    ^~~~~~
nano/af_nano.c:149:14: warning: unused variable 'i' [-Wunused-variable]
     int ano, i;
              ^
nano/af_nano.c:149:9: warning: unused variable 'ano' [-Wunused-variable]
     int ano, i;
         ^~~
nano/af_nano.c: In function 'nano_connect':
nano/af_nano.c:197:22: error: 'sockaddr_na' undeclared (first use in this function)
     if (get_user(&(((sockaddr_na *)uservaddr)->sun_family)) != AF_NANO) {
                      ^~~~~~~~~~~
nano/af_nano.c:197:22: note: each undeclared identifier is reported only once for each function it appears in
nano/af_nano.c:197:35: error: expected expression before ')' token
     if (get_user(&(((sockaddr_na *)uservaddr)->sun_family)) != AF_NANO) {
                                   ^
nano/af_nano.c:209:14: warning: implicit declaration of function 'sock_awaitconn' [-Wimplicit-function-declaration]
     if ((i = sock_awaitconn(sock, serv_upd->npd_socket, flags)) < 0)
              ^~~~~~~~~~~~~~
nano/af_nano.c: In function 'nano_getname':
nano/af_nano.c:292:30: warning: passing argument 1 of 'move_addr_to_user' from incompatible pointer type [-Wincompatible-pointer-types]
     return move_addr_to_user(&upd->npd_sockaddr_na, upd->npd_sockaddr_len,
                              ^
In file included from nano/af_nano.c:12:0:
/home/elks/Downloads/elks/elks/include/linuxmt/socket.h:58:12: note: expected 'char *' but argument is of type 'struct sockaddr_na *'
 extern int move_addr_to_user(char *,size_t,char *,int *);
            ^~~~~~~~~~~~~~~~~
nano/af_nano.c:293:9: warning: passing argument 3 of 'move_addr_to_user' from incompatible pointer type [-Wincompatible-pointer-types]
         usockaddr, usockaddr_len)
         ^~~~~~~~~
In file included from nano/af_nano.c:12:0:
/home/elks/Downloads/elks/elks/include/linuxmt/socket.h:58:12: note: expected 'char *' but argument is of type 'struct sockaddr *'
 extern int move_addr_to_user(char *,size_t,char *,int *);
            ^~~~~~~~~~~~~~~~~
nano/af_nano.c:294:1: error: expected ';' before '}' token
 }
 ^
nano/af_nano.c: At top level:
nano/af_nano.c:460:24: error: unknown type name 'select_table'
          int sel_type, select_table * wait)
                        ^~~~~~~~~~~~
nano/af_nano.c:558:5: error: 'nano_select' undeclared here (not in a function)
     nano_select,
     ^~~~~~~~~~~
make[3]: *** [../Makefile-rules:244: nano/af_nano.o] Error 1
make[3]: Leaving directory '/home/elks/Downloads/elks/elks/net'
make[2]: *** [Makefile:123: net/net.a] Error 2
make[2]: Leaving directory '/home/elks/Downloads/elks/elks'
make[1]: *** [Makefile:73: Image] Error 2
make[1]: Leaving directory '/home/elks/Downloads/elks/elks'
make: *** [Makefile:13: all] Error 2
Build script has terminated with error 5
```

--- a/elks/arch/i86/drivers/char/meta.c
+++ b/elks/arch/i86/drivers/char/meta.c
@@ -3,7 +3,11 @@
  *
  * Copyright (C) 1999 by Alistair Riddoch
  *
- * ELKS meta driver for user space device drivers (UDDs)
+ * ELKS driver for user space device drivers (UDDs)
+ *
+ * FIXME This driver is experimental and does NOT currently work.
+ * FIXME It uses character device MAJOR_NR to index blk_dev[], must be fixed.
+ * FIXME Also requires device parameter to request_fn() routine.
  */
 
 #include <linuxmt/config.h>
@@ -17,15 +21,15 @@
 #include <linuxmt/udd.h>
 #include <linuxmt/debug.h>
 
-#define MAJOR_NR	MISC_MAJOR
-
+#define MAJOR_NR	UDD_MAJOR	/* FIXME "CURRENT" macro fails in end_request()*/
 #define METADISK
-
 #include "../block/blk.h"
 
 struct ud_driver drivers[MAX_UDD];
 
 struct ud_request requests[MAX_UDR];
+
+static int udd_major = -1;	/* FIXME required until device passed to request_fn*/
 
 static struct ud_driver *get_driver(int major)
 {
@@ -70,7 +74,7 @@ int post_request(struct ud_driver *driver, struct ud_request *request)
 static void do_meta_request(void)
 {
     //int major = MAJOR(device);
-    int major = MAJOR(CURRENT->rq_dev); /* FIXME: change MAX_BLKDEV and create /dev/meta*/
+    int major = udd_major; /* FIXME required until device passed to request_fn*/
     struct ud_driver *driver = get_driver(major);
     struct ud_request *udr;
     struct request *req;
@@ -81,20 +85,15 @@ static void do_meta_request(void)
 	end_request(0);
 	return;
     }
-    printk("1");
     while (1) {
 	req = blk_dev[major].current_request;
-	printk("2");
 	if (!req || req->rq_dev < 0 || req->rq_sector == -1)
 	    return;
-	printk("5");
 	udr = new_request();
 	udr->udr_type = UDR_BLK + req->rq_cmd;
 	udr->udr_ptr = req->rq_sector;
 	udr->udr_minor = MINOR(req->rq_dev);
-	printk("6");
 	post_request(driver, udr);
-	printk("7");
 
 	/* Should really check here whether we have a request */
 	if (req->rq_cmd == WRITE) {
@@ -106,7 +105,6 @@ static void do_meta_request(void)
 	    		buff, kernel_ds, 1024/2);
 #endif
 	}
-	printk("8");
 
 	/* Wake up the driver so it can deal with the request */
 	wake_up(&driver->udd_wait);
@@ -248,8 +246,7 @@ void meta_select()
 /* Do nothing */
 }
 
-static int meta_ioctl(struct inode *inode,
-		      struct file *filp, int cmd, char *arg)
+static int meta_ioctl(struct inode *inode, struct file *filp, int cmd, char *arg)
 {
     struct ud_driver *driver;
     int i, minor = MINOR(inode->i_rdev);
@@ -286,12 +283,11 @@ static int meta_ioctl(struct inode *inode,
 	    blk_dev[driver->udd_major].request_fn = do_meta_request;
 	}
 	printk("device\n");
+	udd_major = driver->udd_major;	/* FIXME until device passed to request_fn*/
 	driver->udd_task = current;
-#if 0
-	driver->udd_rwait = NULL;	/* FIXME: Not valid */
-#endif
 	return minor;
-      reg_err:
+
+reg_err:
 	driver->udd_type = UDD_NONE;
 	return -ENODEV;
     }
@@ -304,9 +300,6 @@ static int meta_ioctl(struct inode *inode,
     switch (cmd) {
     case META_POLL:
 	printk("waiting for request\n");
-#if 0
-	driver->udd_wait = NULL;	/* FIXME: not valid */
-#endif
 	driver->udd_req = NULL;
 	wake_up(&driver->udd_rwait);
 	interruptible_sleep_on(&driver->udd_wait);
@@ -329,9 +322,6 @@ static int meta_ioctl(struct inode *inode,
 	verified_memcpy_fromfs(driver->udd_req, arg,
 			       sizeof(struct ud_request_trunc));
 	wake_up(&driver->udd_req->udr_wait);
-#if 0
-	driver->udd_req->udr_wait = NULL;	/* FIXME: Not valid */
-#endif
 	interruptible_sleep_on(&driver->udd_req->udr_wait);
 	break;
     default:
@@ -373,10 +363,8 @@ static struct file_operations meta_chr_fops = {
 
 void meta_init(void)
 {
-    int i;
-
-    printk("meta driver Copyright (C) 1999 Alistair Riddoch\n");
-    if (!(i = register_chrdev(MAJOR_NR, DEVICE_NAME, &meta_chr_fops)))
+    printk("Userspace device driver Copyright (C) 1999 Alistair Riddoch\n");
+    if (!register_chrdev(MAJOR_NR, DEVICE_NAME, &meta_chr_fops))
 	meta_initialised = 1;
 }
 

--- a/elks/arch/i86/drivers/char/meta.c
+++ b/elks/arch/i86/drivers/char/meta.c
@@ -40,7 +40,6 @@ static struct ud_driver *get_driver(int major)
 
 struct ud_request *new_request(void)
 {
-    int i;
     register struct ud_request *r = requests;
 
     do {
@@ -68,9 +67,10 @@ int post_request(struct ud_driver *driver, struct ud_request *request)
 }
 
 
-static void do_meta_request(kdev_t device)
+static void do_meta_request(void)
 {
-    int major = MAJOR(device);
+    //int major = MAJOR(device);
+    int major = MAJOR(CURRENT->rq_dev); /* FIXME: change MAX_BLKDEV and create /dev/meta*/
     struct ud_driver *driver = get_driver(major);
     struct ud_request *udr;
     struct request *req;
@@ -78,7 +78,7 @@ static void do_meta_request(kdev_t device)
 
     printk("do_meta_request %d %x\n", major, blk_dev[major].current_request);
     if (NULL == driver) {
-	end_request(0, req->rq_dev);
+	end_request(0);
 	return;
     }
     printk("1");
@@ -117,7 +117,7 @@ static void do_meta_request(kdev_t device)
 	/* REQUEST HAS BEEN RETURNED BY USER PROGRAM */
 	/* request must be dealt with and ended */
 	if (udr->udr_status == 1) {
-	    end_request(0, req->rq_dev);
+	    end_request(0);
 	    udr->udr_status = 0;
 	    continue;
 	}
@@ -132,7 +132,7 @@ static void do_meta_request(kdev_t device)
 	    		driver->udd_data, driver->udd_task->mm.dseg, 1024/2);
 #endif
 	}
-	end_request(1, req->rq_dev);
+	end_request(1);
 	wake_up(&udr->udr_wait);
     }
 }

--- a/elks/include/linuxmt/major.h
+++ b/elks/include/linuxmt/major.h
@@ -24,7 +24,7 @@
  *  4 - /dev/tty*,ttyp*,ttyS*                         char tty, pty slave, serial
  *  5 -
  *  6 - /dev/lp                /dev/rom               block romflash
- *  7 -
+ *  7 -                        /dev/udd               block meta user device driver
  *  8 - /dev/tcpdev                                   kernel <-> ktcp comm
  *  9 - /dev/eth                                      NIC driver
  * 10 - /dev/cgatext
@@ -40,7 +40,7 @@
 #define TTY_MAJOR         4
 #define TTYAUX_MAJOR      5
 #define LP_MAJOR          6
-#define MISC_MAJOR        7
+#define UDD_MAJOR         7
 #define TCPDEV_MAJOR      8
 #define ETH_MAJOR         9  /* should be rather a network-class driver */
 #define CGATEXT_MAJOR     10

--- a/elks/include/linuxmt/udd.h
+++ b/elks/include/linuxmt/udd.h
@@ -14,8 +14,8 @@
 
 #include <linuxmt/fs.h>
 
-#define MAX_UDD	8
-#define MAX_UDR	32
+#define MAX_UDD	1 /* FIXME only 1 UDD until device passed to request_fn*/
+#define MAX_UDR	8
 
 struct ud_driver {
     int udd_type;

--- a/elks/include/linuxmt/udd.h
+++ b/elks/include/linuxmt/udd.h
@@ -20,7 +20,7 @@
 struct ud_driver {
     int udd_type;
     int udd_major;
-    struct task *udd_task;
+    struct task_struct *udd_task;
     char *udd_data;
     struct ud_request *udd_req;
     struct wait_queue udd_wait;
@@ -37,7 +37,7 @@ struct ud_driver {
 struct ud_driver_trunc {
     int udd_type;
     int udd_major;
-    struct task *udd_task;
+    struct task_struct *udd_task;
     char *udd_data;
 };
 

--- a/elks/net/nano/af_nano.c
+++ b/elks/net/nano/af_nano.c
@@ -18,6 +18,8 @@
 
 #ifdef CONFIG_NANO
 
+int sock_awaitconn(register struct socket *mysock, struct socket *servsock, int flags);
+
 struct nano_proto_data nano_datas[NSOCKETS_NANO];
 
 static struct nano_proto_data *nano_data_alloc(void)
@@ -146,10 +148,8 @@ static int nano_release(register struct socket *sock, struct socket *peer)
 static int nano_bind(struct socket *sock,
 		     struct sockaddr *umyaddr, int sockaddr_len)
 {
-    int ano, i;
     register struct nano_proto_data *upd = sock->data;
     register struct nano_proto_data *pupd;
-    unsigned short old_ds;
 
     printk("nano_bind %x\n", sock);
 
@@ -194,7 +194,7 @@ static int nano_connect(register struct socket *sock,
 /*    if (sock->state == SS_CONNECTED)
 	return -EISCONN;*/	/*Already checked in socket.c*/
 
-    if (get_user(&(((sockaddr_na *)uservaddr)->sun_family)) != AF_NANO) {
+    if (get_user(&(((struct sockaddr_na *)uservaddr)->sun_family)) != AF_NANO) {
 	printk("ADR - {%d}\n", sockna.sun_family);
 	return -EINVAL;
     }
@@ -289,8 +289,8 @@ static int nano_getname(register struct socket *sock,
     } else
 	upd = NA_DATA(sock);
 
-    return move_addr_to_user(&upd->npd_sockaddr_na, upd->npd_sockaddr_len,
-				    usockaddr, usockaddr_len)
+    return move_addr_to_user((char *)&upd->npd_sockaddr_na, upd->npd_sockaddr_len,
+				    (char *)usockaddr, usockaddr_len);
 }
 
 static int nano_read(register struct socket *sock,
@@ -456,8 +456,7 @@ static int nano_write(register struct socket *sock,
     return (size - todo);
 }
 
-static int nano_select(register struct socket *sock,
-		       int sel_type, select_table * wait)
+static int nano_select(register struct socket *sock, int sel_type)
 {
     struct nano_proto_data *upd, *peerupd;
 
@@ -469,10 +468,10 @@ static int nano_select(register struct socket *sock,
 	if (sel_type == SEL_IN) {
 	    if (sock->iconn)
 		return 1;
-	    select_wait(sock->wait, wait);
+	    select_wait(sock->wait);
 	    return (sock->iconn ? 1 : 0);
 	}
-	select_wait(sock->wait, wait);
+	select_wait(sock->wait);
 
 	return 0;
     }
@@ -485,7 +484,7 @@ static int nano_select(register struct socket *sock,
 	else if (sock->state != SS_CONNECTED)
 	    return 1;
 
-	select_wait(sock->wait, wait);
+	select_wait(sock->wait);
 
 	return 0;
     }
@@ -499,7 +498,7 @@ static int nano_select(register struct socket *sock,
 	if (NA_BUF_SPACE(peerupd) > 0)
 	    return 1;
 
-	select_wait(sock->wait, wait);
+	select_wait(sock->wait);
 
 	return 0;
     }

--- a/image/Make.devices
+++ b/image/Make.devices
@@ -170,6 +170,11 @@ endif
 #	$MKSET  0 15 $(MKDEV) /dev/loop c 7
 
 ##############################################################################
+# UDD user device driver, experimental
+
+#	$(MKDEV) /dev/udd c 7 0
+
+##############################################################################
 # TCPDEV, used by ktcp
 
 	$(MKDEV) /dev/tcpdev c 8 0

